packages/expo-updates/ios/EXUpdates/EXUpdatesAppController.m

```diff
@@ -324,6 +378,60 @@ - (void)_emergencyLaunchWithFatalError:(NSError *)error
   }
 }
 
+- (void)_registerObservers
+{
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_handleJavaScriptDidLoad) name:RCTJavaScriptDidLoadNotification object:nil];
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(_handleJavaScriptDidFailToLoad) name:RCTJavaScriptDidFailToLoadNotification object:nil];
+}
+
+- (void)_unregisterObservers
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self name:RCTJavaScriptDidLoadNotification object:nil];
+  [[NSNotificationCenter defaultCenter] removeObserver:self name:RCTJavaScriptDidFailToLoadNotification object:nil];
+}
+
+- (void)_handleJavaScriptDidLoad
+{
+  [self _unregisterObservers];
+  dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 10 * NSEC_PER_SEC), _controllerQueue, ^{
+    [self _unsetRCTErrorHandlers];
+  });
```

Comment:
As we chatted about, let's replace this 10-second window with listeners for discrete phases of the launch lifecycle. This way we have more consistent behavior across devices.

Perhaps it would be helpful to think about the phases of the lifecycle, ex: JS loaded, JS parsed, JS evaluated (all the code in module scope runs, all exports get defined), initial React render runs, initial componentDidMount calls run, initial native views created.

While not a watertight defense against backwards incompatibility, waiting for the initial native views to be created might be a good point in the lifecycle at which to unset the error handlers. It is unlikely the filesystem or SQLite or other persistent storage would be modified in a backwards-incompatible way (but the possibility of it still makes me wary).

Depending on where the failure occurs in the lifecycle we might want to have different policies. For example, if the JS fails to parse, the update is probably not going to run (at least without a new native app update). If the initial React render fails, we want to fall back to another update but we may want to try the failed update under different conditions in the future.

Also, if the initial React render succeeds during one launch but fails during a future one, it seems dangerous to roll back because backwards-incompatible changes may have been made to the filesystem or other persistent storage. This makes me think we need a flag (if we don't have one already) to mark that an update did successfully run (at least far enough to get past the initial render) and that we should not rollback except in specific scenarios (ex: developer marked the update as safe to roll back, or the update's JS failed to parse for some reason).